src/cryptography/x509.py

```diff
@@ -249,6 +264,96 @@ def __repr__(self):
                 "path_length={0.path_length})>").format(self)
 
 
+class GeneralName(object):
+    _gn_str_types = [
+        GeneralNameType.rfc822_name,
+        GeneralNameType.dns_name,
+        GeneralNameType.uniform_resource_identifier,
+    ]
+
+    def __init__(self, type, value):
+        if not isinstance(type, GeneralNameType):
+            raise TypeError("type must be a GeneralNameType")
+
+        if type in [
+            GeneralNameType.other_name,
+            GeneralNameType.x400_address,
+            GeneralNameType.edi_party_name
+        ]:
+            raise ValueError("{0} is not a supported type".format(type))
+
+        if (
+            type in self._gn_str_types and not
+            isinstance(value, six.string_types)
```

Comment:
I'm not sure. @glyph has stated to me in the past that he strongly considers domain names, et al to be strings and not bytes. They are, however, a strict subset of ascii (a-z, A-Z, 0-9, and hyphen). It seems odd to require u"mydomain.com" on Python 2. I chose string_types because it allows `str` on both Py2 and Py3, but you're right that it also allows `b'string'` on Py2 and not on Py3. Maybe the lesser of two evils is to just say you pass unicode (and potentially do charset validation in the future when we allow creation)?